src/System.Linq/src/System/Linq/Distinct.cs

```diff
@@ -116,6 +116,11 @@ public List<TSource> ToList()
                 return FillSet().ToList();
             }
 
+            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                return FillSet().ToHashSet(comparer);
```

Comment:
@JonHanna Maybe in the future System.Linq should just switch to `HashSet`... the perf overhead of removing the no free list optimization seems like it'd only be [a branch per iteration](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/HashSet.cs#L961). Besides, we will have to anyway to optimize things like `Except` / `Intersect`.